models/staircase.py

```diff
@@ -0,0 +1,374 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+
+import argparse
+import math
+from enum import IntEnum
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from models.transformer_seq import (FeedForwardLayer, MultiHeadSeqAttention,
+                                    TransformerOutput)
+from models.utils import pos_emb, skew
+
+
+class VARIANT(IntEnum):
+    # staircase transformers
+    VARIANT_ONE = 1
+
+
+def str2bool(v):
+    if isinstance(v, bool):
+        return v
+    if v.lower() in ("yes", "true", "t", "y", "1"):
+        return True
+    elif v.lower() in ("no", "false", "f", "n", "0"):
+        return False
+    else:
+        raise argparse.ArgumentTypeError("Boolean value expected.")
+
+
+def add_args(parser):
+    parser.add_argument(
+        "--staircase-size",
+        type=int,
+        default=64,
+        help="number of tokens in each transformer forward",
+    )
+    parser.add_argument(
+        "--max-staircase-size-forward",
+        type=int,
+        default=63,
+        help="max number of fresh tokens considered",
+    )
+    parser.add_argument(
+        "--fix-staircase-size-forward",
+        type=int,
+        default=-1,
+        help="max number of fresh tokens considered",
+    )
+    parser.add_argument(
+        "--validation-staircase-size-forward",
+        type=int,
+        default=32,
+        help="max number of fresh tokens considered during validation",
+    )
+    parser.add_argument(
+        "--staircase-split-query",
+        action="store_true",
+        default=False,
+        help="query and key, value are not the same size",
+    )
```

Comment:
It looks like this isn't used in any of the experiment scripts? Perhaps remove arguments that aren't required for reproducibility? (Please feel free to push back if all args are important! I don't have full context on staircase!)